bounded_sum_known_n/BoundedSumKnownN.tex

```diff
@@ -0,0 +1,394 @@
+    \item The Rust code uses the stability relation $\dout \geq \din \cdot (\texttt{U}-\texttt{L})/2$ instead of the relation found in the pseudocode, namely $2\cdot \dout \geq \texttt{inf\_cast}(\din, T)\cdot (\texttt{U}-\texttt{L})$. There is already a PR to resolve this discrepancy: \url{https://github.com/opendp/opendp/pull/315}.
+    \item The Rust code does not check for possible overflow errors when performing the arithmetic operation $\texttt{U}-\texttt{L}$, whereas the current pseudocode includes a \texttt{checked\_sub} for it.
+    \item The Rust code does not include an \texttt{inf\_cast} in the stability relation, which is necessary for type correctness, but there is already a PR that resolves this discrepancy: \url{https://github.com/opendp/opendp/blob/493fe87273adca3294551faa927039b755c1756e/rust/opendp/src/trans/sum/mod.rs#L65-L94}.
+    \item The previous PR also includes the new user-specified type trait \texttt{InfDiv}, which the current pseudocode does not include.
+\end{itemize}
+
+These changes will soon be applied to the Rust code. The pseudocode also uses the variable \texttt{n} instead of Rust's \texttt{size}, which is more conventional for proof writing.
+
+
+\section{Algorithm Implementation}
+\subsection{Code in Rust}
+The current OpenDP library contains the transformation \texttt{make\_bounded\_sum\_n} implementing the bounded sum function with known $n$. This is defined in lines 29-49 of the file \texttt{mod.rs} in the Git repository\footnote{Updated on September 20, 2021.} (\url{https://github.com/opendp/opendp/blob/8bbb0fab1da9b86c50235a36d7026189be43b1ab/rust/opendp/src/trans/sum/mod.rs#L29-L49}).
+
+%\begin{figure}[ht]
+    %\includegraphics[width=15cm]{bounded_sum_2.png}
+    %\centering
+    %\label{fig:code}
+%\end{figure}
+
+\begin{figure}[ht]
+    \includegraphics[width=15cm]{bounded_sum_n_new.png}
+    \centering
+    \label{fig:code}
+\end{figure}
+
+\subsection{Pseudocode in Python}
+\label{sec:pseudocode}
+We present a simplified Python-like pseudocode of the Rust implementation below. The necessary definitions for the pseudocode can be found in Section \ref{sec:versioned-docs}.
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{n} must be of type \texttt{usize}.
+        \item Type \texttt{T} must have traits \texttt{DistanceConstant(IntDistance)}, \texttt{TotalOrd}, \texttt{CheckedMul}, \texttt{CheckNull}, \texttt{Sum(Output=T)}, \texttt{Sub(Output=T)},  and \texttt{ExactIntCast(usize)}.
+        \item \texttt{IntDistance} must have trait \texttt{InfCast(T)}. (Note that this bullet point is not needed in this proof, but it is needed in the code so a hint can be constructed; otherwise a binary search would be needed to construct the hint.)
```

Comment:
Isn't the trait InfCast(T) on IntDistance used in the relation itself to do inf_cast(d_in,T)?

Reply:
No, "IntDistance has trait InfCast(T)" means that a value can be inf_casted from type T to type IntDistance. `inf_cast(d_in,T)` casts from the type of d_in to type T.

In the case of `inf_cast(d_in,T)`, because `T` has trait `DistanceConstant(IntDistance)`, `T` automatically has trait `InfCast(IntDistance)`. The fact that having trait `DistanceConstat(IntDistance)` means a type has trait `InfCast(IntDistance)` should be covered in the definition for trait `DistanceConstant` in the pseudocode definitions document.

Reply:
Ah, I see - I got it backwards.  But if IntDistance has trait InfCast(T) and T has trait InfCast(IntDistance), doesn't this effectively imply that T is the same as IntDistance?  Are there any examples of distinct types where we can exactly cast in both directions?  If not, why not simplify things and simply require T=IntDistance?

Reply:
That's what I thought until last week. However, a type `T` can actually have trait InfCast(U)` if any value `v` of type `U` can **either** be casted to a value `w` of type `T` at least as large as `v` **or** can return an error saying the `inf_cast` failed.

So, for example, `u32` has trait `InfCast(u64)` because there's a definition out there for which values `v` of type `u64` can be casted to a value of `u32` at least as big as `v`, and which values `v` of type `u64` can't be casted to a value of `u32` and should instead return an error.

Reply:
InfCast is fallible. If InfCast can cast to an output number GTE the input number, it will, otherwise it (and the entire relation) fails. This means InfCast is implemented for every combination of numeric types. See the third block in [my type conversion table](https://docs.google.com/spreadsheets/d/1DJohiOI3EVHjwj8g4IEdFZVf7MMyFk_4oaSyjTfkO_0/edit#gid=0).